system/ee/legacy/libraries/Functions.php

```diff
@@ -471,33 +502,79 @@ public function form_declaration($data)
-
-		if ($data['enctype'] == 'multi' OR strtolower($data['enctype']) == 'multipart/form-data')
+        
+		if ($data['onsubmit'] != '')
 		{
-			$data['enctype'] = 'enctype="multipart/form-data" ';
-		}
-
-		$form  = '<form '.$data['id'].$data['class'].$data['name'].'method="post" action="'.$data['action'].'" '.$data['onsubmit'].' '.$data['enctype'].">\n";
+            $data['onsubmit'] = ' onsubmit="'.trim($data['onsubmit']).'"';
+        }
+        
+        // Next section is for the 'pass-through' functionality
+
+        $_tagparams = array();
+        $_pass_thru = '';
+        $valid_html_attributes = array();
+                
+        // $valid_html_attributes is an array of valid HTML attributes that might be specified in a FORM tag (this list has all $deft values removed, and does not include
+        // data-* or aria-* as these are processed differently).  The list taken from https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form on 5/june/20
+
+        $valid_html_attributes = explode('|','accept|accept-charset|accesskey|autocapitalize|autocomplete|contenteditable|contextmenu|dir|draggable|dropzone|exportparts|hidden|inputmode|is|itemid|itemprop|itemref|itemscope|itemtype|lang|method|novalidate|onabort|onautocomplete|onautocompleteerror|onblur|oncancel|oncanplay|oncanplaythrough|onchange|onclick|onclose|oncontextmenu|oncuechange|ondblclick|ondrag|ondragend|ondragenter|ondragexit|ondragleave|ondragover|ondragstart|ondrop|ondurationchange|onemptied|onended|onerror|onfocus|oninput|oninvalid|onkeydown|onkeypress|onkeyup|onload|onloadeddata|onloadedmetadata|onloadstart|onmousedown|onmouseenter|onmouseleave|onmousemove|onmouseout|onmouseover|onmouseup|onmousewheel|onpause|onplay|onplaying|onprogress|onratechange|onreset|onresize|onscroll|onseeked|onseeking|onselect|onshow|onsort|onstalled|onsubmit|onsuspend|ontimeupdate|ontoggle|onvolumechange|onwaiting|part|rel|slot|spellcheck|style|tabindex|target|title|translate|role');
+        
+        // Cannot think of any circumstance where tagparams is not empty, but just in case... 
+        if ( ! empty(ee()->TMPL->tagparams))
+        {
+            $_tagparams = ee()->TMPL->tagparams;
+            
+            foreach ($_tagparams as $key => $val) {
+            
+                // Ignore the parameter if $key is defined in $deft (and so already being processed by function)
+                // or if the parameter is not in the list of approved attributes
+                // or if the parameter begins with either aria- or data-
+                if ((! array_key_exists($key, $deft) && in_array($key, $valid_html_attributes)) || (preg_match("/^data-|^aria-/",$key))) {
+        
+                    // Append the key to end of the $_pass_thru variable
+                    // only add a leading space if the variable already has some content
+                    $_pass_thru .= ($_pass_thru == '' ? $key : ' '.$key);
```

Comment:
Instead of building a string here it would probably be cleaner to add the key/value pair to an array, then at the end of the loop [assemble the string of attributes](https://stackoverflow.com/questions/18081625/how-do-i-map-an-associative-array-to-html-element-attributes#18081767).

Reply:
Not sure I understand why creating an array with these values and then turning that array back into an string prior to appending the string to the other strings would be an improvement.  

The only purpose of the activity is to insert a string into the `<form>` tag that is being assembled from other strings.  We don't have any need to access or reorder the values within the string we are assembling.

Perhaps you could explain what I am missing here. Is there some performance benefit?

Reply:
Using an array is just better programming practices than using a long string like that. It's not a performance issue. It's easier to modify an array to add or subtract elements than it is a string. If unit tests are added it'll be a little easier to write the tests etc

Reply:
Sorry - I simply don't buy the concept.  

Two functional lines later the whole lot is concatenated into a long string. 

So to traverse one line of code we build an array and then unbuild it back into a concatenated string.  I can think of no circumstances where in the processing of that intervening line anyone will need to modify or add / subtract elements.  

We literally don't care what they are: we know that they are legitimate to pass through, and that some may have values set.  

I am intrigued to know what circumstance you have in mind for the modify / add / subtract use case.

With regard to unit testing - as noted elsewhere I don't have any familiarity with what testing regime you are using, and so cannot comment.  But a testing regime that only tests array values doesn't sound right to me.  Bear in mind we are simply collecting up some values and dumping them out into an HTML tag.  I'm fascinated to know what kind of test you have in mind, and how the arrays fit in / make it easier.